src/3d/symbols/qgspointcloud3dsymbol_p.cpp

```diff
@@ -521,4 +521,118 @@ Qt3DRender::QGeometry *QgsRGBPointCloud3DSymbolHandler::makeGeometry( Qt3DCore::
+{
+  QgsPointCloudAttributeCollection attributes;
+  constexpr int xOffset = 0;
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "X" ), QgsPointCloudAttribute::Int32 ) );
+  const int yOffset = attributes.pointRecordSize();
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Y" ), QgsPointCloudAttribute::Int32 ) );
+  const int zOffset = attributes.pointRecordSize();
+  attributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Z" ), QgsPointCloudAttribute::Int32 ) );
+
+  QString attributeName;
+  bool attrIsX = false;
+  bool attrIsY = false;
+  bool attrIsZ = false;
+  QgsPointCloudAttribute::DataType attributeType = QgsPointCloudAttribute::Float;
+  int attributeOffset = 0;
+  QgsClassificationPointCloud3DSymbol *symbol = dynamic_cast<QgsClassificationPointCloud3DSymbol *>( context.symbol() );
+  if ( symbol )
+  {
+    int offset = 0;
+    QgsPointCloudAttributeCollection collection = context.attributes();
+
+    if ( symbol->renderingParameter() == QLatin1String( "X" ) )
+    {
+      attrIsX = true;
+    }
+    else if ( symbol->renderingParameter() == QLatin1String( "Y" ) )
+    {
+      attrIsY = true;
+    }
+    else if ( symbol->renderingParameter() == QLatin1String( "Z" ) )
+    {
+      attrIsZ = true;
+    }
+    else
+    {
+      const QgsPointCloudAttribute *attr = collection.find( symbol->renderingParameter(), offset );
+      if ( attr )
+      {
+        attributeType = attr->type();
+        attributeName = attr->name();
+        attributeOffset = attributes.pointRecordSize();
+        attributes.push_back( *attr );
+      }
+    }
+  }
+
+  if ( attributeName.isEmpty() && !attrIsX && !attrIsY && !attrIsZ )
+    return;
+
+  QgsPointCloudRequest request;
+  request.setAttributes( attributes );
+  std::unique_ptr<QgsPointCloudBlock> block( pc->nodeData( n, request ) );
+  if ( !block )
+    return;
+
+  const char *ptr = block->data();
+  int count = block->pointCount();
+  const std::size_t recordSize = attributes.pointRecordSize();
+
+  const QgsVector3D scale = pc->scale();
+  const QgsVector3D offset = pc->offset();
+
+  for ( int i = 0; i < count; ++i )
+  {
+    qint32 ix = *( qint32 * )( ptr + i * recordSize + xOffset );
+    qint32 iy = *( qint32 * )( ptr + i * recordSize + yOffset );
+    qint32 iz = *( qint32 * )( ptr + i * recordSize + zOffset );
+
+    double x = offset.x() + scale.x() * ix;
+    double y = offset.y() + scale.y() * iy;
+    double z = offset.z() + scale.z() * iz;
+    QVector3D point( x, y, z );
+
+    QgsVector3D p = context.map().mapToWorldCoordinates( point );
+    float iParam = 0.0f;
+    if ( attrIsX )
+      iParam = x;
+    else if ( attrIsY )
+      iParam = y;
+    else if ( attrIsZ )
+      iParam = z;
+    else
+      context.getAttribute( ptr, i * recordSize + attributeOffset, attributeType, iParam );
+
+    if ( context.isFilteredOut( ( int ) iParam ) )
```

Comment:
There's potentially a substantial speed boost you'd get if you instead built an unordered set of the filtered values just outside of this loop (ie around line 598) and tested values here using the set instead. It would certainly be much easier for the compiler to optimise vs the un-inlineable call to isFilteredOut.